T1: Refactor for more pythonic code

dominiquesydow opened the pull request:
## Description

In this PR, I am revisiting our talktorial T1 with the aim to make our code more pythonic (started during RDKit hackathon 2019).

## To Do

### Refactor code
- [x] Refactor convert_to_NM function: Less redundancy (e.g. elifs), add exception for unknown units
- [x] Refactor canonical SMILES extraction (use apply and lambda)
- [x] Refactor IC50 to pIC50 conversion (use apply and lambda)
- [x] Use `inplace=True` instead of overwriting old DataFrame with updated DataFrame
- [x] Update variable names, e.g. `target_P00533` > `targets`, `bioactivities` > `bioactivities_api`, `bioact` > `bioactivities`
- [x] Fix ROMol display of molecules in last DataFrame (currently str not image shown)
    - [x] This should be fixed with latest `rdkit`. Check!
- [x] Remove `convert_to_nM`: We do not have to do this conversion manually based on the `value` and `unit` fields, but can use the `standard_value` and `standard_unit` fields directly! Remove all non-nM entries in `standard_unit`.
- [x] How to fetch data by ChEMBL version? You cannot (see [issue](https://github.com/volkamerlab/TeachOpenCADD/issues/22)). Instead filter fetched bioactivity data by bioactivity IDs from ChEMBL 27 release.

### Do code cosmetics
- [x] Split long lines of code to multiple lines (or even split into multiple commands)
- [x] Sort imports
- [x] Remove spaces before/after = when setting function parameters

### Change text
- [x] Rename headings (heading levels)
- [x] Split text cells to smaller bits
- [x] Unify spelling (dataframe, DataFrame)
- [x] Add more description
- [x] Add Fig. 2 contribution (previously missing)

## Questions
None.

## Status
- [x] Ready to go

dominiquesydow commented:
@morgeral - Could you please review the refactoring of T1 (not time sensitive)? 
There is one issue unresolved yet, i.e. in the last DataFrame, the ROMol column shows strings not the compound structures. Do you know why?
For viewing the changes, you can use the ReviewNB as posted above, thanks!

jaimergp commented on talktorials/1_ChEMBL/T1_ChEMBL.ipynb:
```diff
@@ -10,7 +10,7 @@
     "\n",
```
<p>Do you want to return <code>None</code> here? Either raise an exception or return the unaltered value, I'd say?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

dominiquesydow replied:
<p>Later we use that function in:</p><p> </p><p><code>bioactivities_df.apply(lambda x: convert_to_nM(x['units'], x['value']), axis=1)</code></p><p> </p><p>The unaltered value would be added wrongly to the DataFrame and raising an exception will break the apply function. My workaround therefore was using <code>None</code>. What would you do here?</p>

jaimergp replied:
Fair enough! No problems then.

jaimergp commented on talktorials/1_ChEMBL/T1_ChEMBL.ipynb:
```diff
@@ -10,7 +10,7 @@
     "\n",
```
<p>If you want "aligned" results, check the <a href="https://docs.python.org/3/library/string.html#formatspec" target="_blank">formatting minilanguage</a> for space padding. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

dominiquesydow replied:
<p>Thanks, updated output accordingly.</p><p> </p>

dominiquesydow commented:
Short update on the last open issue:

With the current environment (rdkit 2019.03.4) I cannot see the images of the compounds in the last DataFrame, I only see a string "<img src="data:image/png;base64,iVBORw0KGgoAAA..." (ROMol column).

Under the old environment the images are displayed (I think rdkit 2018.3.4).

I have not found a way yet to make the images appear again. Ideas are very welcome.

jaimergp commented:
Installing everything from scratch to debug this.

jaimergp commented on talktorials/1_ChEMBL/T1_ChEMBL.ipynb:
```diff
@@ -10,7 +10,7 @@
     "\n",
```
<p>Technically is not a list, but a _lazy_ list (generator-like). Data is not there until is requested. That's why the DataFrame construction takes so long some cells below: only when <code>pandas</code> iterates over the "list" is the data being downloaded. If you do <code>bioactivities[0]</code>, the request took place at the precise moment. See that if you do <code>bioactivies[100]</code> it takes longer.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

dominiquesydow replied:
<p>Thanks for pointing this out! I changed this to the following:</p><p> </p><p>The results of the query are stored in <code>bioactivities</code>, a <code>QuerySet</code>, i.e. the results are not fetched from ChEMBL unitl we ask for it (here using <code>pandas.DataFrame.from_records</code>).</p><p> </p><p>More information about the <code>QuerySet</code> datatype:</p><p> </p><p>&gt; QuerySets are lazy – the act of creating a QuerySet doesn’t involve any database activity. You can stack filters together all day long, and Django won’t actually run the query until the QuerySet is evaluated. </p><p>From: https://docs.djangoproject.com/en/3.0/topics/db/queries/#querysets-are-lazy</p>

jaimergp commented on talktorials/1_ChEMBL/T1_ChEMBL.ipynb:
```diff
@@ -10,7 +10,7 @@
     "\n",
```
<p>There are some fields named <code>standard_*</code> that already provide this unit normalization. Is this being discussed in the notebook?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

dominiquesydow replied:
<p>I did not know that. In that case, I would drop the whole conversion part and use <code>standard_value</code>  and <code>standard_unit</code>  instead of <code>value</code>  and <code>unit</code>. Is that ok with you, @jaimergp and @andreavolkamer?</p>

AndreaVolkamer replied:
I agree on using the precaluclated values, but from a teaching point of view, I think it is helpful for the reader to understand how these measurements can be converted, so could we leave a description on this in?

dominiquesydow replied:
It turns out, the `standard_units` allow multiple units:

> Why are there so many different types of Standard Units in the database?
> The published units are taken directly from the literature and we then attempt to standardise these to report as a standard type. This is an ongoing curation task, which was started with the most common units first.

https://chembl.gitbook.io/chembl-interface-documentation/frequently-asked-questions/chembl-data-questions#why-are-there-so-many-different-types-of-standard-units-in-the-database

This is why I find multiple standard units `['nM' 'ug.mL-1' '/uM']`, while I did expect only to find `nM`.

Thus, we will need the conversion step anyways.

Please do correct me, if I am overseeing or misunderstanding something here.

dominiquesydow replied:
I dropped all non-nM entries.

jaimergp commented:
I don't know who is to blame, `pandas` or `rdkit`, but you can workaround the issue by wrapping the dataframe as exported HTML:

```
from IPython.display import display, HTML
display(HTML(output_df.drop("smiles", axis=1).head().to_html()))
```

dominiquesydow commented:
> I don't know who is to blame, `pandas` or `rdkit`, but you can workaround the issue by wrapping the dataframe as exported HTML:
> 
> ```
> from IPython.display import display, HTML
> display(HTML(output_df.drop("smiles", axis=1).head().to_html()))
> ```
> 
> ![image](https://user-images.githubusercontent.com/2559438/67414253-fab2e780-f5c2-11e9-83bf-fcdbeb60b19b.png)

This is working again with the new RDKit version.

dominiquesydow commented:
Hi @AndreaVolkamer, coming back to your comment on standardized bioactivity values:

> I agree on using the precaluclated values, but from a teaching point of view, I think it is helpful for the reader to understand how these measurements can be converted, so could we leave a description on this in?

I added a sentence commenting on this:
![image](https://user-images.githubusercontent.com/7207093/82057306-6ecded80-96c3-11ea-8ebc-f9c80bdd5024.png)

Nevertheless, I removed the conversion function, since we are not using it anymore.

Are you ok with this?

dominiquesydow commented:
@jaimergp and @AndreaVolkamer if you both approve, this branch is ready to be merged (as discussed not with `master` but with a new branch branched from `master`, e.g. `package`.

dominiquesydow commented:
@jaimergp and @AndreaVolkamer, I got a reply from the ChEMBL support team: "At the moment, it’s not possible to extract data from web services (or the interface) for a previous ChEMBL release but we hope to work on this in the future."

Our hack could be that (after fetching the ChEMBL data) we filter the results by pre-defined ChEMBL IDs (that live in a `data/ci_chembl_compound_ids.txt` file). What do you think?

jaimergp commented:
Yes, that might be our only way forward. I suggest gzipping the plain text file in that case.

jaimergp commented:
I got an idea for this. Since I have to download all ChEMBL27 for `kinoml` anyway, I can get all the IDs contained in that release. That should also tie us to a specific version! Question is: which IDs do we need? The activities ids or the compound ids?

dominiquesydow commented:
> I got an idea for this. Since I have to download all ChEMBL27 for `kinoml` anyway, I can get all the IDs contained in that release. That should also tie us to a specific version! Question is: which IDs do we need? The activities ids or the compound ids?

Excellent!

I think, we are on the safe site to save both bioactivity and compound ID. 

In theory, we only need the bioactivity ID (we go from target ID > bioactivity ID > compound ID). 
Once a bioactivity was measured and deposited, the target and compound IDs should not change - unless there was an error made and they fixed it.

One more thought on this: I guess it can happen, that bioactivity and compound IDs that we have in our future list will be deleted in a new release. Thus, we will never have a bullet-proof approach here, right?

jaimergp commented:
I don't think they will just delete entries or modify their contents (e.g. mislabeled ligands). Our approach here should be something like:

1. Do as we were doing so far: query online ChEMBL and get all bioactivities
2. Compare the `activity_id` with the frozen set I have attached here (notebook  included). This is a simple test for membership: `this-activity-id in all_activities_from_chembl_27`.
3. Filter out those that are not in this group.

_If_ they delete things, yes, results will be different. The only robust approach here is to instruct people to download the SQL dump for the corresponding release, but that's overkill. A brief mention on the notebook about this would be enough, IMO.

This zip files include instructions to recreate the activities in future ChEMBL releases, and also how to load the NPY file for quick tests.

[timestamped-chembl-queries.zip](https://github.com/volkamerlab/TeachOpenCADD/files/4771401/timestamped-chembl-queries.zip)

dominiquesydow commented:
Great, thanks a lot! 
I will extend the notebook with the steps indicated by you on Monday, hoping we can close this PR next week :)

jaimergp commented:
Yay! Best spot to check for the IDs would be after the dataframe creation at cell 11 in the [current notebook](https://github.com/volkamerlab/TeachOpenCADD/blob/bdd4eebf9652abe356c60157209afc817835a84d/talktorials/1_ChEMBL/T1_ChEMBL.ipynb).

dominiquesydow commented:
@jaimergp - Monday turned into Wednesday, sorry. The filtering step for ChEMBL 27 bioactivities is added. Would you like to take a look?

jaimergp commented:
I will merge to a branch first, as we did with Pymol!
